### PR TITLE
Support Symbolic Arrays of arbitrary length

### PR DESCRIPTION
**Pull request by ven-k**
## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

Following is now supported:
```julia
@mtkmodel ModelWithArrays begin
    @structural_parameters begin
        N = 2
        M = 3
    end
    @parameters begin
        p1[1:4]
        p2[1:N]
        p3[1:N, 1:M] = 10, [description = "A multi-dimensional array of arbitrary length with description"]
        (p4[1:N, 1:M] = 10), [description = "An alternate syntax for p3 to match the syntax of vanilla parameters macro"]
    end
    @variables begin
        v1(t)[1:2] = 10, [description = "An array of variable `v1`"]
        v2(t)[1:3] = [1, 2, 3]
    end
end
```
A dedicated section, in the docs, describes it.

### CODE REVIEW

**Comment by ven-k** [comment]
The tests pertaining to Model-parsing ([model_parsing](https://github.com/SciML/ModelingToolkit.jl/actions/runs/11249915844/job/31277729527?pr=3107#step:6:907) and in [dq_units](https://github.com/SciML/ModelingToolkit.jl/actions/runs/11249915844/job/31277729527?pr=3107#step:6:896) and [units](https://github.com/SciML/ModelingToolkit.jl/actions/runs/11249915844/job/31277729527?pr=3107#step:6:897))have passed. The failure looks unrelated.

**Comment by ven-k** [comment]
This PR is rebased off the master and NO_VALUE changes are included.

File: test/model_parsing.jl [hunk 1]
```diff
@@ -278,6 +279,21 @@ end
     @test MockModel.structure[:defaults] == Dict(:n => 1.0, :n2 => "g()")
 end
 
+@testset "Arrays using vanilla-@variable syntax" begin
```
Comment (ChrisRackauckas) [inline test/model_parsing.jl:282]: Missing the test using structural parameters

File: docs/src/basics/MTKLanguage.md [hunk 2]
```diff
@@ -324,17 +325,57 @@ For example, the structure of `ModelC` is:
...
+```@example mtkmodel-example
+@mtkmodel ModelWithArrays begin
+    @structural_parameters begin
+        N = 2
+        M = 3
+    end
+    @parameters begin
+        p1[1:4]
+        p2[1:N]
+        p3[1:N, 1:M] = 10,
```
Comment (ChrisRackauckas) [inline docs/src/basics/MTKLanguage.md:354]: is this one substantially more difficult to support?

Comment (ven-k) [reply docs/src/basics/MTKLanguage.md:354]: Not really.
Parser is now more granular. It picks up relevant bits of expression at the parsing step.
I've tried to batch similar ones together ([example](https://github.com/SciML/ModelingToolkit.jl/pull/3107/files#diff-e4267563028a9bba382f312bc5bdff2b5d737c02f2141cfe4b358b1c1396109aR269-R272)) and code is in a that is simpler to maintain.

Actually, let me add doc-strings to each parser step to indicate the syntax it parses.

File: test/model_parsing.jl (hunk 1 above)
Comment (ven-k) [reply test/model_parsing.jl:282]: Added here: https://github.com/SciML/ModelingToolkit.jl/pull/3107/commits/09de19020c99b92d51ba577d13b4bae6b3202d1c#diff-3d89635366ad022b99bf84a74296abeb9423849838c8cfb8b5e8ede09f208c91R294-R313

File: docs/src/basics/MTKLanguage.md (hunk 2 above)
Comment (ven-k) [reply docs/src/basics/MTKLanguage.md:354]: I have added the doc strings matching the block here: https://github.com/SciML/ModelingToolkit.jl/pull/3107/commits/4c0b1edee4f4cae0b05eec9f206e491f372c6955

**Comment by ven-k** [comment]
The doc failure is unrelated (errors stem from initialization.md)

**Pull request merged**